### PR TITLE
Feature/jupyter

### PR DESCRIPTION
**Pull request by junjihashimoto**
This modification can run jupyter.
But hasktorch's module is not loaded. (I continue debugging it.)

```
$ nix-shell -A jupyterEnvironment ./default.nix
[nix-shell:~/git/hasktorch]$ jupyter lab                                                                                   
```

### CODE REVIEW

File: nix/shared.nix [hunk 1]
```diff
@@ -10,13 +10,35 @@ let
...
     inherit (libtorch_src pkgsOld)
       libtorch_cpu
       libtorch_cudatoolkit_9_2
       libtorch_cudatoolkit_10_1
     ;
 
+    jupyterEnvironment =
+      let iHaskell = (jupyter pkgsNew).kernels.iHaskellWith {
+            name = "haskell";
+            packages = p: with p; [ hvega formatting pkgs.haskell.packages."${compiler}".hasktorch_cpu ];
```
Comment (tscholak) [inline nix/shared.nix]:
```suggestion
            packages = p: with p; [ hvega formatting pkgsNew.haskell.packages."${compiler}".hasktorch_cpu ];
```
I think

Comment (junjihashimoto) [reply nix/shared.nix]: Thx!

**Comment by tscholak** [comment]
current issue:
```

14008
  This derivation is not meant to be built, aborting
14009
  
14010
  builder for '/nix/store/l2pqvkc8g9x136b2yzvrmwaicfvr5jxl-jupyterlab-shell.drv' failed with exit code 1
14011
  error: build of '/nix/store/l2pqvkc8g9x136b2yzvrmwaicfvr5jxl-jupyterlab-shell.drv' failed
14012
  ##[error]Action failed with error: Error: The process 'nix-build' failed with exit code 100
14013
  (node:3903) UnhandledPromiseRejectionWarning: Error: The process 'nix-build' failed with exit code 100
14014
      at ExecState._setResult (/home/runner/work/_actions/cachix/cachix-action/v2/node_modules/@actions/exec/lib/toolrunner.js:548:25)
14015
      at ExecState.CheckComplete (/home/runner/work/_actions/cachix/cachix-action/v2/node_modules/@actions/exec/lib/toolrunner.js:531:18)
14016
      at ChildProcess.<anonymous> (/home/runner/work/_actions/cachix/cachix-action/v2/node_modules/@actions/exec/lib/toolrunner.js:431:27)
14017
      at ChildProcess.emit (events.js:200:13)
14018
      at maybeClose (internal/child_process.js:1021:16)
14019
      at Socket.<anonymous> (internal/child_process.js:430:11)
14020
      at Socket.emit (events.js:200:13)
14021
      at Pipe.<anonymous> (net.js:586:12)
14022
  (node:3903) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 1)
14023
  (node:3903) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
14024
  ##[error]Node run failed with exit code 1
```
 ¯\_(ツ)_/¯

**Comment by junjihashimoto** [comment]
@tscholak 
jupeter's settings allow us just using nix-shell.

**Comment by junjihashimoto** [comment]
I will use this way .
https://gist.github.com/MMesch/45256a435c9f78158445214a56527f68
This gist does not support macos.
We should update the hash, when we use mac.
https://github.com/tweag/jupyterWith/pull/77

**Comment by junjihashimoto** [comment]
CI takes 1.5hours to build hasktorch with jupyter.
nix's cache does not work.

File: .github/workflows/nix-jupyterlab.yml [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+    - uses: cachix/install-nix-action@v6
+    - name: Setup repo
+      run: |
+        git submodule init && git submodule update
+    - uses: cachix/cachix-action@v2
+      with:
+        name: hasktorch
+        signingKey: '${{ secrets.CACHIX_SIGNING_KEY }}'
+        file: experimental/jupyterlab/default.nix
+        nixBuildArgs: --option sandbox false
```
Comment (junjihashimoto) [inline .github/workflows/nix-jupyterlab.yml]: This pr pushes the caches of both jupyter and hasktorch to reduce compilation time.

Comment (junjihashimoto) [reply .github/workflows/nix-jupyterlab.yml]: https://github.com/hasktorch/hasktorch/pull/257/checks?check_run_id=476444227#step:5:15245
Why sandbox=false doesn't work?

Comment (junjihashimoto) [reply .github/workflows/nix-jupyterlab.yml]: Oh, nixBuildArgs is cachix/cachix-action@v4's function.

**Comment by junjihashimoto** [comment]
cachix-action@v4 does not call nix-build. It's weird.

**Comment by junjihashimoto** [comment]
https://github.com/cachix/cachix-action/issues/30

**Comment by junjihashimoto** [comment]
It is ready to merge.

**Pull request merged**